### PR TITLE
Add cached receiver

### LINKED ISSUE #259: Utilize Preferred Location to reuse pre-fetch queue across batches
In Spark Streaming and Structured Streaming, partitions are consistently assigned to the same Executor. With this feature, we should make sure we maintain pre-fetch queues across batches (instead of creating and destroying them per batch).

### PR DESCRIPTION
**Pull request by sabeegrewal**
This PR adds the cached receiver and cleans up of the client APIs. 

The `CachedEventHubsReceiver` borrows a connection from the `ClientConnectionPool` and opens a receiver link. This receiver link will stay open until: 
* The Spark job is ended
* An executor is added or removed. 

The average of the last five batch sizes is used as the prefetchCount. 

Close #259

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+      event = i.iterator.next
+      assert(requestSeqNo == event.getSystemProperties.getSequenceNumber,
+             errWrongSeqNo(requestSeqNo, event.getSystemProperties.getSequenceNumber))
+    }
+
+    logAndUpdatePrefetch(batchSize)
+    event
+  }
+
+  private def logAndUpdatePrefetch(batchSize: Int): Unit = {
```
Comment (SreeramGarlapati) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: lets updatePrefetch to latest BatchSize to keep the code simple and behavior predictable.

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+private[spark] object CachedEventHubsReceiver extends CachedReceiver with Logging {
+
+  private def notInitializedMessage(nAndP: NameAndPartition): String = {
+    s"No receiver found for EventHubs ${nAndP.ehName} on partition ${nAndP.partitionId}"
+  }
+
+  type MutableMap[A, B] = scala.collection.mutable.HashMap[A, B]
+
+  private[this] val receivers = new MutableMap[NameAndPartition, CachedEventHubsReceiver]()
```
Comment (sabeegrewal) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: Key should include namespace

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+private class CachedEventHubsReceiver(ehConf: EventHubsConf, nAndP: NameAndPartition)
+    extends Logging {
+
+  import org.apache.spark.eventhubs._
+
+  val prefetchCounts = new FixedList[Int](5)
+
+  private var _client: EventHubClient = _
+  private def client: EventHubClient = {
```
[the remark this answers is not in the document]
Comment (sabeegrewal) [reply core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: Why? There should only be one thread in this path, right?

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala (hunk 2 above)
Comment (sabeegrewal) [reply core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: Will address in #305

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (SreeramGarlapati) [reply core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: my bad

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  type MutableMap[A, B] = scala.collection.mutable.HashMap[A, B]
+
+  private[this] val receivers = new MutableMap[NameAndPartition, CachedEventHubsReceiver]()
+
+  private def isInitialized(nAndP: NameAndPartition): Boolean = receivers.synchronized {
+    receivers.get(nAndP).isDefined
+  }
+
+  private def get(nAndP: NameAndPartition): CachedEventHubsReceiver = receivers.synchronized {
+    logInfo(s"lookup on $nAndP")
```
Comment (SreeramGarlapati) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: is this logging for every receive call ?

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      val message = notInitializedMessage(nAndP)
+      throw new IllegalStateException(message)
+    })
+  }
+
+  private[eventhubs] override def receive(ehConf: EventHubsConf,
+                                          nAndP: NameAndPartition,
+                                          requestSeqNo: SequenceNumber,
+                                          batchSize: Int): EventData = {
+    receivers.synchronized {
```
Comment (SreeramGarlapati) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: nit: feel free to ignore.. 
one small Idea after our discussion: 
```
val receiver = createIfNotExist(...) // <- this captures all of the isInitialized receiver.update, get into 1 synchronized block ...
receiver.receive()
```

Comment (SreeramGarlapati) [reply core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]:
```
override def receive(ehConf: EventHubsConf,
                       nAndP: NameAndPartition,
                       requestSeqNo: SequenceNumber,
                       batchSize: Int): EventData = {

    val receiver = createCachedReceiverIfNotExist(ehConf, nAndP, requestSeqNo)
    receiver.receive(requestSeqNo, batchSize)
  }

  def createCachedReceiverIfNotExist(ehConf: EventHubsConf,
                                     nAndP: NameAndPartition,
                                     requestSeqNo: SequenceNumber) : CachedEventHubsReceiver = {
    var receiver: CachedEventHubsReceiver = null
    receivers.synchronized {
      receiver = receivers.getOrElseUpdate(nAndP, new CachedEventHubsReceiver(ehConf, nAndP)) // take care to initialize this lazily
    }

    receiver
  }
```

**Pull request merged**